zebra-consensus/src/chain.rs

```diff
@@ -57,30 +72,30 @@ where
     type Future =
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        match (self.checkpoint.poll_ready(cx), self.block.poll_ready(cx)) {
-            // First, fail if either service fails.
-            (Poll::Ready(Err(e)), _) => Poll::Ready(Err(VerifyChainError::Checkpoint(e))),
-            (_, Poll::Ready(Err(e))) => Poll::Ready(Err(VerifyChainError::Block(e))),
-            // Second, we're unready if either service is unready.
-            (Poll::Pending, _) | (_, Poll::Pending) => Poll::Pending,
-            // Finally, we're ready if both services are ready and OK.
-            (Poll::Ready(Ok(())), Poll::Ready(Ok(()))) => Poll::Ready(Ok(())),
-        }
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Correctness:
+        //
+        // We can't call `poll_ready` on the block and checkpoint verifiers here,
+        // because each `poll_ready` must be followed by a `call`, and we don't
+        // know which verifier we're going to choose yet.
+        // See #1593 for details.
+        Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, block: Arc<Block>) -> Self::Future {
         match block.coinbase_height() {
             Some(height) if height <= self.max_checkpoint_height => self
-                .checkpoint
-                .call(block)
+                .checkpoint_verifier
+                .clone()
+                .oneshot(block)
                 .map_err(VerifyChainError::Checkpoint)
                 .boxed(),
             // This also covers blocks with no height, which the block verifier
             // will reject immediately.
             _ => self
-                .block
-                .call(block)
+                .block_verifier
+                .clone()
+                .oneshot(block)
```

Comment:
These services are now `Buffer`ed, so we don't need an `async` block. We can just call `ServiceExt::oneshot` on a cloned service.

Reply:
@yaahc @oxarbitrage the best way to avoid `poll_ready` / `call` mismatches is to just use a `oneshot`.

It handles the `poll_ready` and `call` internally, so we can't possibly get it wrong.